S3 methods for DataFrame and LazyFrame: simplest

vincentarelbundock opened the pull request:
Issue https://github.com/pola-rs/r-polars/issues/104

Notes:

* S3 methods have `@noRd` tags because I did not want to pollute the manual. But for some reason skipping `nrow` and `ncol` led `check()` to issue a warning, so I included minimal docs for just those two.

TODO in a different PR:

  - [ ] brackets – lots of work\!
  - [ ] is.null (new method needed)
  - [ ] na.omit (new method needed)
  - [ ] colnames (new method needed)

This PR implements a few of the main base R S3 methods for `DataFrame` and `LazyFrame` objects. For example:

``` r
library(rpolars)
d = pl$DataFrame(mtcars)
dl = pl$DataFrame(mtcars)$lazy()

dim(d)
# [1] 32 11

head(dl, 3)
# [1] "polars LazyFrame naive plan: (run ldf$describe_optimized_plan() to see the optimized plan)"
#   SLICE[offset: 0, len: 3]
#     DF ["mpg", "cyl", "disp", "hp"]; PROJECT */11 COLUMNS; SELECTION: "None"

names(d)
#  [1] "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"   "gear"
# [11] "carb"

nrow(dl)
# [1] 32

ncol(dl)
# [1] 11

as.matrix(d) |> str()
#  num [1:32, 1:11] 21 21 22.8 21.4 18.7 18.1 14.3 24.4 22.8 19.2 ...
#  - attr(*, "dimnames")=List of 2
#   ..$ : NULL
#   ..$ : chr [1:11] "mpg" "cyl" "disp" "hp" ...
```

sorhawell commented:
>implement attributes and methods instead of LazyFrame()$collect()

I think e.g. nrow cannot be inferred without collect, however it could be implemented something like where only the count is returned. If polars will optimize away if possible.

```R
x$count()$collect()
```

eitsupi commented on vignettes/polars.Rmd:
````diff
@@ -51,88 +59,106 @@ dependencies other than R itself. This makes it very fast and lightweight to
 # R
-df$select(pl$col(c("A", "B"))$sum())
+df$groupby("id")$mean()
 ```
````
Could you please split the code block between Python and R and set the info string?

like

````rmd
```python
df.method
```

```r
df$method
```
````

vincentarelbundock replied:
d5c256fdfa359a403bcce662a7cec7ff04b1de24

vincentarelbundock commented:
I have completed all I planned to achieve in this PR, so feel free to take a look when you find some time. As always, I'm happy to make any changes you'd like.

The cool thing about my two recent PRs is that it allowed me to considerably simplify the Get Started vignette (@grantmcdermott may want to take a look). Here's a rendered version for your convenience:

https://arelbundock.com/polars.html

Here are the main things I changed in the vignette:

* Added a couple examples of standard base R functions like `head(df)`.
* The first Polars methods users now encounter are the ultra-simple and "clean" ones that we just merged, like `dat$tail()` and  `dat$mean()`.
* Introduce methods chaining using those same ultra-simple methods: `dat$tail(10)$mean()`. I used this example to replace the old Hello World. It was cute and funny, but overly complicated. I must admit that the first time I saw that I had a pretty bad reaction to the ugly syntax. In editing, I thought it was best for the first examples to be simple and actually useful.
* Push the discussion of data types to the end. This is fundamental and useful, but "Get Started" is also a "sales pitch" and I wanted to give users a payoff as fast as possible.

sorhawell commented:
> For what it’s worth, if we didn’t care about overwriting the columns, the previous query could have been written more concisely as:

"For what it’s worth, the previous query could have been written more concisely as"
```R
dat = pl$DataFrame(mtcars)
dat$with_columns(
  pl$col(c("mpg", "hp"))$sum()$over("cyl")$prefix("sum_")
)
```

#BTW also possible but not fully stabilized, so do not recommend in tutorial

```R

# has currenyly a thread-safety user warning. But could be removed now.
dat$with_columns(
  pl$col(c("mpg", "hp"))$sum()$over("cyl")$map_alias(\(x) paste0("sum_",x))
)

# has not stabilized becaues does not work well with wild-cards like $all()
pl$set_polars_options(named_exprs = TRUE) # must be activted like this first
dat$with_columns(
  sum_mpg = pl$col("mpg")$sum()$over("cyl"),
  sum_hp  = pl$col("hp")$sum()$over("cyl")
)


#could be fun to try allow this also :)  as an opt-in feature. All column names converted to `pl$col(x)` and in scope of the method. Not sure it would work well for lazy API though.

dat$with_columns(
  sum_mpg = mpg$sum()$over(cyl),
  sum_hp  = hp$sum()$over(cyl)
)
```

sorhawell commented:
Half-way through the [revised tutorial ](https://arelbundock.com/polars.html). It has really been a pleasure to read. I agree with your changes. I will add a few syntax edits.

grantmcdermott commented:
> Half-way through the [revised tutorial ](https://arelbundock.com/polars.html). It has really been a pleasure to read. I agree with your changes. I will add a few syntax edits.

Agree, lots to like here. Will leave a few comments for you to think about.

grantmcdermott commented on vignettes/polars.Rmd:
```diff
@@ -51,88 +56,105 @@ dependencies other than R itself. This makes it very fast and lightweight to
-and/or regular R vectors.
+## Methods and pipelines
+
+Although some simple R functions work out of the box on **polars** objects, the full power of Polars is realized via _methods_. Polars methods are accessed using the `$` syntax. For example, to convert Polars `Series` and `DataFrames` back to standard R objects, we use the `$to_r_vector()` and `$as_data_frame()` methods:
```
It's been a while since I tried this, but is `as_data_frame` still a tibble-exported (or maybe arrow-exported) function? Just worth thinking ahead to whether users might run into a NAMESPACE clash.

vincentarelbundock replied:
There's no clash between exported functions and polars methods, though, right? And this PR adds a proper `as.data.frame()` s3 for polars objects, which just calls `df$as_data_frame()` under the hood.

grantmcdermott replied:
Cool cool. Just wanted to check.

grantmcdermott commented on vignettes/polars.Rmd:
````diff
@@ -51,88 +56,105 @@ dependencies other than R itself. This makes it very fast and lightweight to
+```python
 # Python
-df.select([pl.col(["A", "B"]).sum()]))
+df.group_by("id").mean()
+```
 
+```r
 # R
-df$select(pl$col(c("A", "B"))$sum())
+df$groupby("id")$mean()
````
Not saying you should revert, but my motivation for including two columns in the original example was to show the users would also need to switch over to R's native `c()` method (i.e., in addition to switching to `$`).

vincentarelbundock replied:
Right that's useful info. But given that we haven't shown the user *anything* yet and it's their very first contact with **polars**, I feel that absolute simplicity and clarity serves our pitch better at this point. It's also not too hard to figure out `c()` in time when users try it out. That said, I don't feel extremely strongly about this, and could be swayed.

vincentarelbundock replied:
Also, there is an example with `dat$select(pl$col(c("mpg", "hp")))` a little further down the page. So I think that's enough.

grantmcdermott commented on vignettes/polars.Rmd:
````diff
@@ -390,3 +441,21 @@ pl$DataFrame(iris)$select(
+pl$dtypes$Float64
+```
+
+The full list of valid Polars types can be found by typing `pl$dtypes` 
+into your R console. These include _Boolean_, _Float32(64)_, _Int32(64)_,
+_Utf8_, _Categorical_, _Date_, etc. Note that some type names differ from what
+they are called in R (e.g., _Boolean_ in Polars is equivalent to `logical()` in
+R). This might occasionally require you to look up a specific type. But the good
+news is that **polars** generally does a good job of inferring types
+automatically.
+
````
On balance, I agree that moving types to the end makes sense for an intro vignette. But we've already included a type coercion above via `cast`. So it might be an idea to either flag it above, or at least mention something about type coercion here again.

grantmcdermott reviewed (commented):
Lots to like here @vincentarelbundock.

One thing I haven't mentioned as a specific comment---but I think is good for consistency---is sticking to the hard 80 character wrapping. I know that you're a fellow vim user and [this is the hook](https://vim.fandom.com/wiki/Format_only_long_lines) I always use ;-)

vincentarelbundock commented on vignettes/polars.Rmd:
```diff
@@ -17,12 +17,28 @@ options(rmarkdown.html_vignette.check_title = FALSE)
-written in Rust and can be used directly from via the polars crate. Orginally Polars bindings were implemented for Python. The **polars** R-package also provides equivalent bindings from R. Within
-R, python and rust each polars package is just called polars. The implemention polars in the
-different languages can be called rust-polars, py-polars, r-polars, nodejs-polars ...
+Polars is [a lightning fast](https://h2oai.github.io/db-benchmark/) Data Frame
```
I will update the link. And wow, DuckDB has really improved! Do you think we should change the 5-10 improvement relative to dplyr in light of those results? Some operations are 50x faster...

grantmcdermott replied:
Good question. We could hedge it with a classic "orders of magnitude(s) improvement".

vincentarelbundock replied:
Good call.

vincentarelbundock commented:
FYI, I believe I have addressed all the points raised in review so far. Feel free to make more requests if needed.

sorhawell commented:
Many thanks @vincentarelbundock  :)
